### ISSUE TITLE
Data.IOArray.newArray runtime exception with negative length

### ISSUE THREAD

**Issue by c-snail** [idris-lang/Idris2#885]
`newArray` function from `Data.IOArray` raises a runtime exception if provided a negative length.

# Steps to Reproduce
```idris
import Data.IOArray

main : IO ()
main = () <$ newArray {elem=Unit} (-10)
```

# Expected Behavior
Program fails to type check
Or newArray returns IO (Maybe (IOArray ...))
Or when provided with a negative length, newArray acts as if it were provided `0`

# Observed Behavior
Runtime exception
```
Exception in make-vector: -10 is not a valid vector length
```
with chez-scheme backend
```
RangeError: Invalid array length
```
with javascript backend run in node

**Comment by SmiVan** [comment]
Failing to typecheck sounds like the most appropriate one of the expected behaviors.

At the moment there are no checks for the size value when calling newArray.
Maybe importing Data.So and using `So (size >= 0)` could work?

**Comment by buzden** [comment]
Is there any reason of allowing to create an array of size zero?

**Comment by SmiVan** [comment]
I've tried implementing this but making `So (size >= 0)` work is a mountain of trouble it turns out.

Maybe Nat could be used for array sizes instead?
Or would the performance loss be too much of a problem.

**Comment by buzden** [comment]
`Nat`s are potentially infinite unlike `Int`s (and thus treated at runtime like `Integer`s) and I guess this was the only reason to use `Int`s in a place where `Nat`s fit naturally. Maybe the performance loss worth benchmarking to decide maybe to switch to `Nat`s finally.

However, is won't solve the problem of the similar (but worse) problem of indexing white reading or writing. `Nat`s there are hardly better than `Int`s because the value is limited from both sides. It's natural work for `Fin` datatype since `maxSize` field exists in `IOArray` record (i.e. to have `writeArray : HasIO io => (ar : IOArray elem) -> Fin ar.maxSize -> elem -> io ()` once `maxSize` field is `Nat`).

But as far as I know `Fin`s don't have a special runtime treatment like `Nat`s have, and this can lead to severe performance loss (I wish to be wrong here ;-) ).

**Comment by stefan-hoeck** [comment]
I think what is actually needed here is a proper order type for `Int`. Together with `Data.DPair.Subset`, such a thing could be used both for storing the size as well as for indexing. Luckily, an order type has been added recently by gallais (albeit to contrib): `Data.Int.Order` with proper treatment of transitivity and other laws. (This is, however not yet included in contrib.ipkg, so I submitted  #1041).

Since this is such an important thing to have (in terms of performance AND type safety), wouldn't it make sense to move `Data.Int.Order` to base and use that one here:

```idris
import Data.Int.Order
import public Data.DPair

NonNeg : Type
NonNeg = Subset Int (LTE 0)

FinInt : Int -> Type
FinInt size = Subset Int (\n => (LTE 0 n, LT n size))
```
We could use `NonNeg` for the size of arrays and for initialization, and `FinInt` for indexing.
Of course this would need some cleanup plus some performant utilities for conveniently converting `Int`s to `NonNeg` and `FinInt`. But it might give us the best in terms of safety and performance here.

**Comment by melted** [comment]
Just note that this is used in the compiler, so changing the API will lead to bootstrap problems. It's probably best to add a new safer way to create buffers as a separate function.

**Comment by melted** [comment]
> Is there any reason of allowing to create an array of size zero?

I was just thinking about this, and yes, size should be allowed to be zero. It's a trivial case but very useful. If you make buffers out of lists of something and write them to a file you don't want to have to implement special handling for empty lists.

**Comment by gallais** [comment]
I had to use `IOArray` today and the interface is pretty unsatisfactory.
* You have to initialise with `Nothing` even if you have a better default (e.g. `0`)
* You have to perform bound checks even if you know you're safe (e.g. iterating over `[0..w-1]` on a `w`-wide array)
* Failure is silent (out of bound write just returns `()`)

**Comment by SmiVan** [comment]
Maybe a new IOArray-like module should be created, so the design can be improved while avoiding bootstrap problems?

**Comment by gallais** [comment]
We need to think about designing a nice `interface` and we can then have different
representations that can be used seamlessly.

Because even with a new `ArrayIO`, it's still super dubious to use an `Array Bool`
(no matter whether that's just `Bool` or `Maybe Bool` under the hood) instead of an
`Array Bits64` that can be 64 times smaller and store the same information.

**Comment by Z-snails** [comment]
For special cases like `Bool` it would be reasonable to create specialised functions, but it general it would be very hard, because the memory representation varies widely vetween backends, so it would be entirely backend dependant (and anyway `Bool` is turned into an `Int`, so it wouldn't help). *

One idea I've had for creating `IOArray`s more easily is a few functions like:
```idris
unfold : (elem -> elem) -> (size : Int) -> elem -> IOArray elem

fromIndex : (Fin size -> elem) -> (size : Nat) -> IOArray elem
```
This would get rid of the need for so many `Maybe`s (and reduce the memory use/number of indirections)

*In fact, packing would be almost impossible in my GRIN backend because the memory representation depends on the context of the program.

**Comment by gallais** [comment]
> For special cases like `Bool` it would be reasonable to create specialised functions

I agree but I have good hope that if you can write a routine explaining how to store 64 `Bool`s into a
`Bits64` then you should get a lifting to any number of `Bool`s in an `Array Bits64`.

And the same automated lifting should get you the ability to store arbitrarily many `Bits8` into an
`Array Bits64` by explaining how you fit 8 of them into a single `Bits64`.

This is what I got tonight. `loop` was surprisingly hard to write because of `Array`'s `size` parameter
until I realised I needed to capture the whole `write` action.

```idris
module Data.Array

import Data.DPair
import Data.Nat
import Data.Vect

%default total

public export
interface Array size arr elt | arr, elt where
  init' : HasIO io => Maybe elt -> io arr
  read  : HasIO io => arr -> Subset Nat (`LT` size) -> io elt
  write : HasIO io => arr -> Subset Nat (`LT` size) -> elt -> io ()

export
init : HasIO io => (n : Nat) -> Array n arr elt => Maybe elt -> io arr
init n = init'

export
tabulate : HasIO io => (n : Nat) -> Array n arr elt =>
           (Subset Nat (`LT` n) -> elt) -> io arr
tabulate n f = do
  vs <- init n (the (Maybe elt) Nothing)
  loop n (\ i => write vs i (f i))
  pure vs

  where
    zero : Subset Nat (`LT` S _)
    zero = Element Z ltZero

    succ : Subset Nat (`LT` bd) -> Subset Nat (`LT` S bd)
    succ (Element k prf) = Element (S k) (LTESucc prf)

    loop : (n : Nat) -> (Subset Nat (`LT` n) -> io ()) -> io ()
    loop Z act = pure ()
    loop (S n) act = do act zero; loop n (act . succ)
```

The idea being that we should then be able to write implementations for:

```idris
Array 64 Bits64 Bool
Array 8 Bits64 Bits8
{m, n : Nat} -> Array n Bits64 elt => Array m (IOArray Bits64) elt
```

**Comment by Z-snails** [comment]
That looks good, I'm planning on having a go at adding optimising any `Nat` like types over the next few days, so we could use `Fin`.

We could split it into two interfaces:
```idris
interface Unboxed maxNum elt | elt where
    IsValid : Fin maxNum -> Bits64 -> Type
    validBOrValid : b1 `LT` b2 => IsValid n1 b1 -> IsValid n2 b2 -> (IsValid n1 (b1 .|. b2), IsValid n2 (b1 .|. b2) 
    unbox : elt -> (ind : Fin maxNum) -> Subset Bits64 (IsValid ind)
    box : {auto 0 prf : IsValid ind b} -> (b : Bits64) -> (ind : Fin maxNum) -> elt

interface Unboxed elt => UArray (arr : Nat -> Type) elt | arr where
    read : HasIO io => Fin size -> arr size -> io elt
    write : HasIO io => Fin size -> arr size -> elt -> io ()
```
Because then it's clearer what each one is supposed to do.
Also your single `Array` only supports in-place writes which is a bit inconvenient for unboxing purposes because you then need to worry about `IORef Bits64`.
The `Unboxed` interface might turn out really hard to write because `Bits64` is primitive and we're only looking at a certain subset of the bits.

**Comment by gallais** [comment]
> Also your single `Array` only supports in-place writes which is a bit inconvenient for
  unboxing purposes because you then need to worry about `IORef Bits64`.

Now that I've tried to add an implementation (for `Bits32` because `Bits64` is not yet
supported by `Data.Bits`, cf. #983) I understand what you mean by this. The definition
is straightforward but `write`'s type is indeed annoying unless you wrap the `Bits32` in
an IORef:

```idris
(lte32 : LTE size 32) => Array size Bits32 Bool where
  init' (Just True) = pure oneBits
  init' _ = pure zeroBits

  read arr (Element i ltSize)
    = let 0 bound : LT i 32 = LTEIsTransitive _ _ _ ltSize lte32 in
      pure (testBit arr (Element i bound))

  write arr (Element i ltSize) elt
    = let 0 bound : LT i 32 = LTEIsTransitive _ _ _ ltSize lte32
          act : Bits32 -> Subset Nat (`LT` 32) -> io ()
          act =  if elt then setBit else clearBit
      in
      pure $ act arr (Element i bound) -- you got a `Bits32` back. Now what?!
```

**Comment by gallais** [comment]
Got this working: https://github.com/gallais/Idris2/blob/70c3e0fc766adf6bee9f27d5d042a7cc613a9f08/libs/contrib/Data/Array.idr
